### ISSUE TITLE
Add a caching mechanism to our decoder during generation

### ISSUE THREAD

**Issue by mattdangerw** [keras-team/keras-nlp#705]
It is standard practice to to cache `key` and `value` tensors when generating text, to avoid running the same computations over and over for the static parts of a generated sequence.

See for example:
https://discuss.huggingface.co/t/what-is-the-purpose-of-use-cache-in-decoder/958
https://www.tensorflow.org/api_docs/python/tfm/nlp/models/TransformerDecoder#call

It also appears this caching can be [XLA compatible](https://blog.tensorflow.org/2022/11/how-hugging-face-improved-text-generation-performance-with-xla.html).

We should definitely scope this out for our new generative offerings. How much performance are we leaving on the table without any caching? How feasible will caching be with our current functional model approach?

**Comment by chenmoneygithub** [comment]
so there are two caches mentioned in the discussion thread:
1. Only do the forward pass for the newly generated token, say we just generated Nth token, we only compute the logits for it.
2. Generate the same token if seeing the same leading tokens.

1 could bring us a big speed boost, while I am not sure about 2, since we have certain randomness during generation, should we actually go for generating the same thing every time?

**Comment by chenmoneygithub** [comment]
See CachedAttention implementation here: [code link](https://github.com/tensorflow/models/blob/fc0a5072212ba9b4b3b44a31b0f0845bbe153b0c/official/nlp/modeling/layers/attention.py#L26)

We have some work to do - -"

**Comment by mattdangerw** [comment]
This has been added.
